index.d.ts

```diff
@@ -0,0 +1,44 @@
+declare module 'apollo-datasource-mongodb' {
+  import { DataSource } from 'apollo-datasource'
+  import { Collection as MongoCollection, ObjectId } from 'mongodb'
+  import {
+    Collection as MongooseCollection,
+    Document,
+    Model as MongooseModel
+  } from 'mongoose'
+
+  export type Collection<T> = T extends Document
+    ? MongooseCollection
+    : MongoCollection<T>
+
+  export type Model<T> = T extends Document ? MongooseModel<T> : never
+
+  export type ModelOrCollection<T> = T extends Document
+    ? Model<T>
+    : Collection<T>
+
+  export interface Options {
+    ttl: number
+  }
+
+  export class MongoDataSource<TData, TContext = any> extends DataSource<
+    TContext
+  > {
+    protected collection: Collection<TData>
+    protected model: Model<TData>
```

Comment:
this is only sometimes defined—should it be `model?` ?

Reply:
If we change this to `model?`, then in a project which uses `mongoose`, devs would have to do `this.model?.someMethod` or `this.model && this.model.someMethod`.

As seen on [line 14](https://github.com/GraphQLGuide/apollo-datasource-mongodb/pull/13/files#diff-b52768974e6bc0faccb7d4b75b162c99R14), this is going to conditionally have type `MongooseModel<T>` (if the constructor is specialized with a mongoose collection) or `never` if the constructor is specialized with a mongo collection)

Maybe I should change line 14 to this:

```
  export type Model<T> = T extends Document ? MongooseModel<T> : undefined
```

Reply:
After this change ^

this.model will have type:
- `MongooseModel<T>` if we're using a `MongooseCollection`
- `undefined` if we're using a `MongoCollection`

What do you think?

Reply:
sounds more accurate than never? whatever you think best